apps/dcos_dns/src/dcos_dns_router.erl

```diff
@@ -37,43 +38,45 @@ default_resolvers() ->
     lists:map(fun validate_upstream/1, Resolvers).
 
 %% @private
--spec(find_upstream(Labels :: [binary()]) -> [upstream()] | internal).
-find_upstream([<<"mesos">>|_]) ->
-    dcos_dns_config:mesos_resolvers();
-find_upstream([<<"localhost">>|_]) ->
-    internal;
-find_upstream([<<"zk">>|_]) ->
-    internal;
-find_upstream([<<"spartan">>|_]) ->
-    internal;
-find_upstream([<<"directory">>, <<"thisdcos">>|_]) ->
-    internal;
-find_upstream([<<"global">>, <<"thisdcos">>|_]) ->
-    internal;
-find_upstream([<<"directory">>, <<"dcos">>|_]) ->
-    internal;
-find_upstream(Labels) ->
+-spec(find_upstream_zone(Labels :: [binary()]) -> forward()).
+find_upstream_zone([<<"mesos">>|_]) ->
+   {dcos_dns_config:mesos_resolvers(), <<"mesos.">>};
+find_upstream_zone([<<"localhost">>|_]) ->
+    {internal, <<"localhost.">>};
+find_upstream_zone([<<"zk">>|_]) ->
+    {internal, <<"zk.">>};
+find_upstream_zone([<<"spartan">>|_]) ->
+    {internal, <<"spartan.">>};
+find_upstream_zone([<<"directory">>, <<"thisdcos">>, Zone |_]) ->
+    % care about the cripto id
+    {internal, <<Zone/binary, ".thisdcos.directory.">>};
+find_upstream_zone([<<"global">>, <<"thisdcos">>, Zone |_]) ->
+    {internal, <<Zone/binary, ".thisdcos.global.">>};
+find_upstream_zone([<<"directory">>, <<"dcos">>, Zone |_]) ->
+    {internal, <<Zone/binary, ".dcos.directory.">>};
+find_upstream_zone(Labels) ->
     case find_custom_upstream(Labels) of
-        [] ->
-            default_resolvers();
-        Resolvers ->
+        {[], _ZoneLabels} ->
+            {default_resolvers(), <<".">>};
+        {Resolvers, Zone} ->
             lager:debug("resolving ~p with custom upstream: ~p", [Labels, Resolvers]),
-            Resolvers
+            {Resolvers, Zone}
     end.
 
--spec(find_custom_upstream(Labels :: [binary()]) -> [upstream()]).
+-spec(find_custom_upstream(Labels :: [binary()]) -> forward()).
 find_custom_upstream(QueryLabels) ->
     ForwardZones = dcos_dns_config:forward_zones(),
     UpstreamFilter = upstream_filter_fun(QueryLabels),
-    maps:fold(UpstreamFilter, [], ForwardZones).
+    maps:fold(UpstreamFilter, {[], <<".">>}, ForwardZones).
 
 -spec(upstream_filter_fun([dns:labels()]) ->
-    fun(([dns:labels()], upstream(), [upstream()]) -> [upstream()])).
+    fun(([dns:labels()], upstream(), [upstream()]) -> forward())).
 upstream_filter_fun(QueryLabels) ->
-    fun(Labels, Upstream, Acc) ->
-        case lists:prefix(Labels, QueryLabels) of
+    fun(ZoneLabels, Upstream, Acc) ->
+        case lists:prefix(ZoneLabels, QueryLabels) of
             true ->
-                Upstream;
+                Zone = dcos_net_utils:string_join(ZoneLabels, <<".">>),
```

Comment:
I know this may end up running more than needed, but to move it up would require changes on the return types. Since I am not a big fan of optimizing something that is not the bottleneck, I left this way.